File: expense-tax-management/services/ai-worker/src/ai_worker/enrichment.py
"""Pure deterministic enrichment evaluator.

No network, database, Temporal, or Foundry imports. Given a validated
ExpenseEnrichmentInputV1 it returns a fully-formed ExpenseEnrichmentResultV1
with outcome=applied, rule-derived tag keys, and history-inferred suggestions.

Historical inference rules (per cross-task spec):
- At least 3 examples (exampleCount >= 3).
- Leading candidate ratio: count / exampleCount >= 0.80.
- No tie at the leading count (second-highest count < leading count).
- Candidate must be eligible (present in the corresponding eligible set).
- Emits source="historical", confidence=matchCount/exampleCount.
- Evidence hash: SHA-256 over {"kind","candidate","exampleCount","matchCount"}
  JSON (keys sorted, no spaces, ASCII).
- Personal expenses (eligibleTaxSnapshot=None) produce no tax suggestion.
- Tax suggestion carries the full snapshot + input expenseVersion.
"""

from __future__ import annotations

import hashlib
import json
from uuid import UUID

from expense_contracts.generated import (
    ExpenseEnrichmentInputV1,
    ExpenseEnrichmentResultV1,
)
from expense_contracts.generated.expense_enrichment_input_v1 import (
    EligibleTaxSnapshot,
    History,
)
from expense_contracts.generated.expense_enrichment_result_v1 import (
    AggregateCounts,
    RuleTagKey,
    Suggestions,
    Suggestions1,
    Suggestions2,
)

from ai_worker.constants import ENRICHMENT_RULES_VERSION

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

_WEEKEND_ISOWEEKDAYS = frozenset({6, 7})  # Saturday=6, Sunday=7


# ---------------------------------------------------------------------------
# Evidence hash
# ---------------------------------------------------------------------------


def _evidence_hash(
    kind: str, candidate: str, example_count: int, match_count: int
) -> str:
    """Canonical SHA-256 over the aggregate evidence tuple."""
    payload = json.dumps(
        {
            "kind": kind,
            "candidate": candidate,
            "exampleCount": example_count,
            "matchCount": match_count,
        },
        separators=(",", ":"),
        sort_keys=True,
    )
    return hashlib.sha256(payload.encode()).hexdigest()


# ---------------------------------------------------------------------------
# Rule engine
# ---------------------------------------------------------------------------


def _deterministic_rule_tags(inp: ExpenseEnrichmentInputV1) -> list[str]:
    """Return eligible rule-derived tag keys for this expense."""
    eligible = {e.root for e in inp.eligibleTagKeys}
    tags: list[str] = []

    # Rule: merchant
    if inp.normalizedMerchant is not None:
        candidate = f"merchant:{inp.normalizedMerchant}"
        if candidate in eligible:
            tags.append(candidate)

    # Rule: weekend
    if (
        inp.incurredOn.isoweekday() in _WEEKEND_ISOWEEKDAYS
        and "timing:weekend" in eligible
    ):
        tags.append("timing:weekend")

    # Rule: selected_category
    if inp.spendingCategoryId is not None:
        candidate = f"category:{inp.spendingCategoryId}"
        if candidate in eligible:
            tags.append(candidate)

    return tags


# ---------------------------------------------------------------------------
# Historical inference helpers
# ---------------------------------------------------------------------------


def _leading_candidate[T](
    candidates: list[tuple[T, int]],
    example_count: int,
) -> tuple[T, int] | None:
    """
    Return (candidate, count) if:
    - example_count >= 3
    - leading count / example_count >= 0.80
    - no tie at the leading count
    Otherwise return None.
    """
    if example_count < 3:
        return None
    if not candidates:
        return None

    sorted_cands = sorted(candidates, key=lambda x: x[1], reverse=True)
    top_val, top_count = sorted_cands[0]

    # Ratio threshold
    if top_count / example_count < 0.80:
        return None

    # Tie check: second-highest must be strictly less
    if len(sorted_cands) > 1 and sorted_cands[1][1] == top_count:
        return None

    return top_val, top_count


def _infer_tags(
    history: History, eligible_tag_keys: frozenset[str]
) -> list[Suggestions]:
    candidates = [
        (ck.key, ck.count)
        for ck in history.candidateTagKeys
        if ck.key in eligible_tag_keys
    ]
    result = _leading_candidate(candidates, history.exampleCount)
    if result is None:
        return []
    tag_key, match_count = result
    return [
        Suggestions(
            kind="tag",
            source="historical",
            tagKey=tag_key,
            confidence=match_count / history.exampleCount,
            aggregateCounts=AggregateCounts(
                exampleCount=history.exampleCount,
                matchCount=match_count,
            ),
            evidenceHash=_evidence_hash(
                "tag", tag_key, history.exampleCount, match_count
            ),
        )
    ]


def _infer_spending_categories(
    history: History, eligible_cat_ids: frozenset[str]
) -> list[Suggestions1]:
    candidates = [
        (str(cc.id), cc.count)
        for cc in history.candidateSpendingCategoryIds
        if str(cc.id) in eligible_cat_ids
    ]
    result = _leading_candidate(candidates, history.exampleCount)
    if result is None:
        return []
    cat_id_str, match_count = result
    return [
        Suggestions1(
            kind="spending_category",
            source="historical",
            spendingCategoryId=UUID(cat_id_str),
            confidence=match_count / history.exampleCount,
            aggregateCounts=AggregateCounts(
                exampleCount=history.exampleCount,
                matchCount=match_count,
            ),
            evidenceHash=_evidence_hash(
                "spending_category", cat_id_str, history.exampleCount, match_count
            ),
        )
    ]


def _infer_tax_categories(
    history: History,
    eligible_tax_snapshot: EligibleTaxSnapshot,
    expense_version: int,
) -> list[Suggestions2]:
    active_tax_cat_ids = frozenset(
        str(tc.root) for tc in eligible_tax_snapshot.activeTaxCategoryIds
    )
    candidates = [
        (str(tc.id), tc.count)
        for tc in history.candidateTaxCategoryIds
        if str(tc.id) in active_tax_cat_ids
    ]
    result = _leading_candidate(candidates, history.exampleCount)
    if result is None:
        return []
    tax_cat_id_str, match_count = result
    return [
        Suggestions2(
            kind="tax_category",
            source="historical",
            taxCategoryDefinitionId=UUID(tax_cat_id_str),
            businessTaxProfileId=eligible_tax_snapshot.businessTaxProfileId,
            businessTaxProfileVersion=eligible_tax_snapshot.businessTaxProfileVersion,
            taxonomyVersionId=eligible_tax_snapshot.taxonomyVersionId,
            taxYear=eligible_tax_snapshot.taxYear,
            expenseVersion=expense_version,
            confidence=match_count / history.exampleCount,
            aggregateCounts=AggregateCounts(
                exampleCount=history.exampleCount,
                matchCount=match_count,
            ),
            evidenceHash=_evidence_hash(
                "tax_category", tax_cat_id_str, history.exampleCount, match_count
            ),
        )
    ]


def _infer_history(
    history: History,
    eligible_tag_keys: frozenset[str],
    eligible_spending_category_ids: frozenset[str],
    eligible_tax_snapshot: EligibleTaxSnapshot | None,
    expense_version: int,
) -> list[Suggestions | Suggestions1 | Suggestions2]:
    suggestions: list[Suggestions | Suggestions1 | Suggestions2] = []
    suggestions.extend(_infer_tags(history, eligible_tag_keys))
    suggestions.extend(
        _infer_spending_categories(history, eligible_spending_category_ids)
    )
    if eligible_tax_snapshot is not None:
        suggestions.extend(
            _infer_tax_categories(history, eligible_tax_snapshot, expense_version)
        )
    return suggestions


# ---------------------------------------------------------------------------
# Public entry point
# ---------------------------------------------------------------------------


def evaluate(inp: ExpenseEnrichmentInputV1) -> ExpenseEnrichmentResultV1:
    """Evaluate enrichment rules and history for a single expense.

    Pure function: no I/O, no randomness, no side effects.
    Always returns outcome=applied; stale/skipped are upstream branches.
    """
    tag_strs = _deterministic_rule_tags(inp)

    eligible_tag_keys = frozenset(e.root for e in inp.eligibleTagKeys)
    eligible_cat_ids = frozenset(str(e.root) for e in inp.eligibleSpendingCategoryIds)

    suggestions = _infer_history(
        history=inp.history,
        eligible_tag_keys=eligible_tag_keys,
        eligible_spending_category_ids=eligible_cat_ids,
        eligible_tax_snapshot=inp.eligibleTaxSnapshot,
        expense_version=inp.expenseVersion,
    )

    return ExpenseEnrichmentResultV1(
        schemaVersion=1,
        rulesVersion=ENRICHMENT_RULES_VERSION,
        outcome="applied",
        ruleTagKeys=[RuleTagKey(k) for k in tag_strs],
        suggestions=suggestions,
    )

File: expense-tax-management/services/workflow-worker/src/activities/enrichment.ts
import { createHash } from "node:crypto";

import {
  ExpenseEnrichmentResultV1Schema,
  type EnrichmentSuggestionResult,
  type ExpenseEnrichmentInputV1,
  type ExpenseEnrichmentResultV1,
} from "@expense-tax/contracts";

function evidenceHash(kind: string, candidate: string, exampleCount: number, matchCount: number): string {
  return createHash("sha256")
    .update(JSON.stringify({ candidate, exampleCount, kind, matchCount }).replace(
      /[\u0080-\uffff]/g,
      (character) => `\\u${character.charCodeAt(0).toString(16).padStart(4, "0")}`,
    ))
    .digest("hex");
}

function leading<T>(candidates: readonly { candidate: T; count: number }[], examples: number) {
  if (examples < 3 || candidates.length === 0) return null;
  const sorted = [...candidates].sort((a, b) => b.count - a.count);
  const first = sorted[0];
  if (!first || first.count / examples < 0.8 || sorted[1]?.count === first.count) return null;
  return first;
}

export function evaluateEnrichment(input: ExpenseEnrichmentInputV1): ExpenseEnrichmentResultV1 {
  const eligibleTags = new Set(input.eligibleTagKeys);
  const rules: string[] = [];
  if (input.normalizedMerchant !== null) {
    const merchant = `merchant:${input.normalizedMerchant}`;
    if (eligibleTags.has(merchant)) rules.push(merchant);
  }
  if ([0, 6].includes(new Date(`${input.incurredOn}T00:00:00.000Z`).getUTCDay()) && eligibleTags.has("timing:weekend")) {
    rules.push("timing:weekend");
  }
  if (input.spendingCategoryId !== null) {
    const category = `category:${input.spendingCategoryId}`;
    if (eligibleTags.has(category)) rules.push(category);
  }

  const { history } = input;
  const examples = history.exampleCount;
  const suggestions: EnrichmentSuggestionResult[] = [];
  const tag = leading(
    history.candidateTagKeys.filter((item) => eligibleTags.has(item.key)).map((item) => ({ candidate: item.key, count: item.count })),
    examples,
  );
  if (tag) suggestions.push({
    kind: "tag", source: "historical", tagKey: tag.candidate,
    confidence: tag.count / examples,
    aggregateCounts: { exampleCount: examples, matchCount: tag.count },
    evidenceHash: evidenceHash("tag", tag.candidate, examples, tag.count),
  });

  const eligibleCategories = new Set(input.eligibleSpendingCategoryIds);
  const category = leading(
    history.candidateSpendingCategoryIds.filter((item) => eligibleCategories.has(item.id)).map((item) => ({ candidate: item.id, count: item.count })),
    examples,
  );
  if (category) suggestions.push({
    kind: "spending_category", source: "historical", spendingCategoryId: category.candidate,
    confidence: category.count / examples,
    aggregateCounts: { exampleCount: examples, matchCount: category.count },
    evidenceHash: evidenceHash("spending_category", category.candidate, examples, category.count),
  });

  const snapshot = input.eligibleTaxSnapshot;
  if (snapshot !== null) {
    const eligibleTaxes = new Set(snapshot.activeTaxCategoryIds);
    const tax = leading(
      history.candidateTaxCategoryIds.filter((item) => eligibleTaxes.has(item.id)).map((item) => ({ candidate: item.id, count: item.count })),
      examples,
    );
    if (tax) suggestions.push({
      kind: "tax_category", source: "historical", taxCategoryDefinitionId: tax.candidate,
      businessTaxProfileId: snapshot.businessTaxProfileId,
      businessTaxProfileVersion: snapshot.businessTaxProfileVersion,
      taxonomyVersionId: snapshot.taxonomyVersionId,
      taxYear: snapshot.taxYear,
      expenseVersion: input.expenseVersion,
      confidence: tax.count / examples,
      aggregateCounts: { exampleCount: examples, matchCount: tax.count },
      evidenceHash: evidenceHash("tax_category", tax.candidate, examples, tax.count),
    });
  }

  return ExpenseEnrichmentResultV1Schema.parse({
    schemaVersion: 1,
    rulesVersion: 1,
    outcome: "applied",
    ruleTagKeys: rules,
    suggestions,
  });
}
